src/rust/src/http_reader.rs

```diff
@@ -34,8 +44,8 @@ pub(crate) fn from_http_err(error: HttpError) -> GeozeroError {
     }
 }
 
-impl HttpFgbReader {
-    pub async fn open(url: &str) -> Result<HttpFgbReader> {
+impl HttpFgbReader<Initial> {
```

Comment:
I've never quite seen this pattern before! It's a nice way to move run-time checks to compile time without having a bunch of boilerplate types like:

```
struct HttpFgbReaderInner {
     client: BufferedHttpRangeClient,
     /// Current read offset
     pos: usize,
 @@ -25,6 +27,14 @@ pub struct HttpFgbReader {
     item_filter: Option<Vec<packed_r_tree::SearchResultItem>>,
     /// Current position in item_filter
     feat_no: usize,
 }

struct InitialHttpFgbReader(HttpFgbReaderInner);
struct OpenedHttpFgbReader(HttpFgbReaderInner);
struct FeatureSelectedHttpFgbReader(HttpFgbReaderInner);

impl InitialHttpFgbReader {
    pub async fn open(url: &str) -> Result<OpenHttpFgbReader> {
        ...
    }
}
```

Nice work!

Reply:
Quite proud of that commit :smile:  But credit goes to https://nostarch.com/rust-rustaceans